pybind/utils/pose.cpp

```diff
@@ -0,0 +1,134 @@
+#include "mplib/utils/pose.h"
+
+#include <memory>
+#include <stdexcept>
+
+#include <pybind11/eigen.h>
+#include <pybind11/numpy.h>
+#include <pybind11/operators.h>
+#include <pybind11/pybind11.h>
+
+#include "docstring/utils/pose.h"
+#include "pybind_macros.hpp"
+
+namespace py = pybind11;
+
+namespace mplib {
+
+using Matrix4 = Eigen::Matrix<S, 4, 4, Eigen::RowMajor>;
+using pyarray_t = py::array_t<S, py::array::c_style | py::array::forcecast>;
+
+void build_utils_pose(py::module &pymp) {
+  auto PyPose =
+      py::class_<Pose<S>, std::shared_ptr<Pose<S>>>(pymp, "Pose", DOC(mplib, Pose));
+
+  // Enable implicit conversions on the Python side from any py::object to Pose<S>
+  py::implicitly_convertible<py::object, Pose<S>>();
+
+  PyPose.def(py::init<>(), DOC(mplib, Pose, Pose))
+      .def(py::init<const Vector3<S> &, const Vector4<S> &>(),
+           py::arg("p") = Vector3<S> {0.0, 0.0, 0.0},
+           py::arg("q") = Vector4<S> {1.0, 0.0, 0.0, 0.0}, DOC(mplib, Pose, Pose, 2))
+      .def(py::init([](const Matrix4 &mat) {
+             return Pose<S> {mat.block<3, 1>(0, 3),
+                             Quaternion<S>(mat.block<3, 3>(0, 0))};
+           }),
+           py::arg("matrix"), DOC(mplib, Pose, Pose, 5))
+      .def(py::init([](const py::object &obj) {
```

Comment:
Maybe add a comment saying that this is for Sapien pose. Also, can you please remind me of what the number at the end of the `DOC` thingy is? For example the 5 here
> DOC(mplib, Pose, Pose, 5)

Reply:
That number selects between different docstrings of overloaded methods.
If you check the docstring, I've mentioned the sapien Pose there. 

> Constructs a Pose with given Python object that has ``p`` and ``q`` attributes
   (e.g., ``sapien.Pose``) or a 4x4 np.ndarray transformation matrix.
   :param obj: a Pose-like object with ``p`` and ``q`` attributes or
         a 4x4 np.ndarray transformation matrix)doc";